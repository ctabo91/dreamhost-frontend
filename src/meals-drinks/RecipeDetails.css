.RecipeDetails-overlay {
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    background-color: rgba(0, 0, 0, 0.6);
    z-index: 1;
    width: 100vw;
    min-height: 100vh;
}

.RecipeDetails-content {
    position: relative;
    z-index: 2;
    color: rgb(255, 255, 255);
}

.RecipeDetails-title {
    /* font-family: "Great Vibes", cursive;
    font-weight: 400;
    font-style: normal; */
    font-family: "Alumni Sans Pinstripe", sans-serif;
    font-weight: 400;
    font-style: normal;
    font-size: 7rem;
}

.RecipeDetails-category {
    font-family: "Great Vibes", cursive;
    font-weight: 400;
    font-style: normal;
    font-size: 2.5rem;
    /* font-family: "Playball", cursive;
    font-weight: 400;
    font-style: normal; */
}

.heading-title {
    font-family: "Cinzel Decorative", serif;
    font-weight: 400;
    font-style: normal;
}

.RecipeDetails-instructions {
    word-wrap: break-word;
    overflow-wrap: break-word;
    white-space: pre-wrap;
    max-height: 400px;
    overflow-y: hidden;
    text-align: justify;
}

.RecipeDetails-instructions:hover {
    overflow-y: auto;
}

.RecipeDetails-instructions::-webkit-scrollbar {
    width: 2px;
}

.RecipeDetails-instructions::-webkit-scrollbar-thumb {
    background-color: gray;
    border-radius: 10px;
}

.btn-favorite {
    position: absolute;
    top: 0;
    right: 0;
    font-size: 2.5rem;
    background-color: transparent;
    border: none;
    color: #fff;
}

.btn-favorite.favorited {
    color: #dc3545;
}

.btn-edit {
    width: 130px;
    font-size: larger;
}

.split-line {
    border-right: 1px solid #ccc;
    margin-left: 10px;
}