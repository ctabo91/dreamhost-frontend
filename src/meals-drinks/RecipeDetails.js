import React, { useState, useEffect, useContext } from "react";
import { useParams, Link } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faHeart as solidHeart } from "@fortawesome/free-solid-svg-icons";
import { faHeart as regularHeart } from "@fortawesome/free-regular-svg-icons";
import DreamHostApi from "../api/api";
import LoadingSpinner from "../common/LoadingSpinner";
import UserContext from "../auth/UserContext";
import "./RecipeDetails.css";

/** Detail page for recipe, showing the name, category, area(or type and glass for drinks), ingredients, and instructions.
 * 
 * - If the access to this recipe is global, a heart icon is shown to favorite the recipe.
 * - If access is personal, an edit button is shown which redirects to an update form for the specific recipe.
 */ 

function RecipeDetails() {
    const { access, item, id } = useParams();
    const { currentUser, hasFavoritedRecipe, favoriteRecipe } = useContext(UserContext);

    const [recipe, setRecipe] = useState(null);

    /** Get a specific recipe by "id" and "item"(meals or drinks). 
     * 
     * If access is personal, get a personal recipe by "id" and "item"(meals or drinks).
     */
    useEffect(() => {
        const getRecipe = async () => {
            access === "personal"
                    ? setRecipe(await DreamHostApi.getPersonalRecipe(currentUser.username, id, item))
                    : setRecipe(await DreamHostApi.getRecipe(id, item));
        }

        getRecipe();
    }, [id, item, currentUser.username, access]);

    const organizeInstructions = (instructionString) => {
        // Check if the string starts with a number followed by a dot and space
        const isNumbered = /^\d+\.\s+/.test(instructionString);
      
        if (isNumbered) {
          // Split the string based on the numbered pattern
          return instructionString.split(/\d+\.\s+/).filter(Boolean);
        } else {
          // Split the string based on sentences
          return instructionString.split('.').map(instruction => instruction.trim()).filter(Boolean);
        }
    }

    const organizedInstructions = recipe ? organizeInstructions(recipe.instructions) : [];

    if (!recipe) return <LoadingSpinner />;

    return (
        <div className="RecipeDetails-container"
             style={{
                backgroundImage: `url(${recipe.thumbnail})`,
                backgroundSize: 'cover',
                backgroundPosition: 'center',
                minHeight: '100vh',
                width: '100%',
                paddingTop: '60px',
                margin: '0',
                position: 'relative',
             }}
        >
            <div className="RecipeDetails-overlay"></div>
            <div className="RecipeDetails-content container">
                <div className="row">
                    <div className="col">
                        <h2 className="RecipeDetails-title text-center mt-4">{recipe.name.toUpperCase()}</h2>
                        <h3 className="RecipeDetails-category text-center">
                            {item === "meals"
                                ? `${recipe.area} ${recipe.category} Dish`
                                : recipe.category}
                        </h3>
                        {item === "drinks" && (
                            <>
                                <h4 className="RecipeDetails-type heading-title text-center mt-3">{recipe.type}</h4>
                                <h6 className="RecipeDetails-glass heading-title text-center">*Serve in a {recipe.glass}*</h6>
                            </>
                        )}
                    </div>
                </div>
                <div className="row mt-4 d-flex justify-content-around">
                    <div className="col-12 col-sm-6">
                        <h4 className="instructions-title heading-title">Instructions</h4>
                        <ol className="RecipeDetails-instructions">
                            {organizedInstructions.map((instruction, idx) => (
                                <li key={idx} className="RecipeDetails-list-item">{instruction}</li>
                            ))}
                        </ol>
                    </div>
                    <div className="col-12 col-sm-1 split-line"></div>
                    <div className="col-12 col-sm-4">
                        <h4 className="ingredients-title heading-title">Ingredients</h4>
                        <ul className="RecipeDetails-ingredients">
                            {recipe.ingredients.map((ingredient, idx) => (
                                <li key={idx} className="RecipeDetails-list-item">{ingredient}</li>
                            ))}
                        </ul>
                    </div>
                </div>
                {access === "global" ? (
                    <button
                        className={`btn btn-favorite font-weight-bold  ${hasFavoritedRecipe(id, item) ? "favorited" : ""}`}
                        onClick={() => favoriteRecipe(id, item)}
                    >
                        <FontAwesomeIcon icon={hasFavoritedRecipe(id, item) ? solidHeart : regularHeart} />
                    </button>
                ) : (
                    <div className="container d-flex justify-content-end">
                        <Link to={`/recipes/update/${item}/${id}`}>
                            <button className="btn btn-edit btn-danger active mt-4 mb-4">Edit Recipe</button>
                        </Link>
                    </div>
                )}
            </div>
        </div>
    );
}


export default RecipeDetails;