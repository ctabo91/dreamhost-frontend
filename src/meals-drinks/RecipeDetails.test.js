import React from 'react';
import { render, screen, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import RecipeDetails from './RecipeDetails';
import UserContext from '../auth/UserContext';
import DreamHostApi from '../api/api';

jest.mock('../api/api');

const mockUserContext = {
  currentUser: {
    username: 'testUser',
  },
  hasFavoritedRecipe: jest.fn(),
  favoriteRecipe: jest.fn(),
};

const mockRecipe = {
  id: 1,
  name: 'Test Recipe',
  area: 'Test Area',
  category: 'Test Category',
  type: 'Test Type',
  glass: 'Test Glass',
  instructions: '1. Test instruction 1. Test instruction 2.',
  ingredients: ['Ingredient 1', 'Ingredient 2'],
  thumbnail: 'test-thumbnail.jpg',
};

beforeEach(() => {
  jest.clearAllMocks();
});

describe('RecipeDetails Component', () => {
  it('matches snapshot', async () => {
    DreamHostApi.getRecipe.mockResolvedValueOnce(mockRecipe);

    const asFragment = await act(async () => {
      const { asFragment } = render(
        <UserContext.Provider value={mockUserContext}>
          <MemoryRouter initialEntries={['/recipes/global/meals/1']}>
            <Routes>
              <Route path="/recipes/:access/:item/:id" element={<RecipeDetails />} />
            </Routes>
          </MemoryRouter>
        </UserContext.Provider>
      );
      return asFragment;
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders loading spinner when recipe is null', async () => {
    await act(async () => {
      render(
        <MemoryRouter>
          <UserContext.Provider value={mockUserContext}>
            <RecipeDetails />
          </UserContext.Provider>
        </MemoryRouter>
      );
    });

    expect(screen.getByText('Loading ...')).toBeInTheDocument();
  });

  it('renders recipe details for meals when item is "meals"', async () => {
    DreamHostApi.getRecipe.mockResolvedValueOnce(mockRecipe);

    await act(async () => {
      render(
        <UserContext.Provider value={mockUserContext}>
          <MemoryRouter initialEntries={['/recipes/global/meals/1']}>
            <Routes>
              <Route path="/recipes/:access/:item/:id" element={<RecipeDetails />} />
            </Routes>
          </MemoryRouter>
        </UserContext.Provider>
      );
    });
    
    expect(screen.getByText('TEST RECIPE')).toBeInTheDocument();
    expect(screen.getByText('Test Area Test Category Dish')).toBeInTheDocument();
    expect(screen.getByText('Test instruction')).toBeInTheDocument();
    expect(screen.getByText('Test instruction 2.')).toBeInTheDocument();
    expect(screen.getByText('Ingredient 1')).toBeInTheDocument();
    expect(screen.getByText('Ingredient 2')).toBeInTheDocument();
  });

  it('renders recipe details for drinks when item is "drinks"', async () => {
    const drinkRecipe = { ...mockRecipe, item: 'drinks' };
    DreamHostApi.getRecipe.mockResolvedValueOnce(drinkRecipe);

    await act(async () => {
      render(
        <UserContext.Provider value={mockUserContext}>
          <MemoryRouter initialEntries={['/recipes/global/drinks/1']}>
            <Routes>
              <Route path="/recipes/:access/:item/:id" element={<RecipeDetails />} />
            </Routes>
          </MemoryRouter>
        </UserContext.Provider>
      );
    });

    expect(screen.getByText('TEST RECIPE')).toBeInTheDocument();
    expect(screen.getByText('Test Category')).toBeInTheDocument();
    expect(screen.getByText('Test Type')).toBeInTheDocument();
    expect(screen.getByText('*Serve in a Test Glass*')).toBeInTheDocument();
  });

  it('renders favorite button when access is "global"', async () => {
    DreamHostApi.getRecipe.mockResolvedValueOnce(mockRecipe);
    mockUserContext.hasFavoritedRecipe.mockReturnValueOnce(false);

    await act(async () => {
      render(
        <UserContext.Provider value={mockUserContext}>
          <MemoryRouter initialEntries={['/recipes/global/meals/1']}>
            <Routes>
              <Route path="/recipes/:access/:item/:id" element={<RecipeDetails />} />
            </Routes>
          </MemoryRouter>
        </UserContext.Provider>
      );
    });

    const favoriteButton = screen.getByRole('button', { class: 'btn-favorite' });
    expect(favoriteButton).toBeInTheDocument();

    userEvent.click(favoriteButton);
    expect(mockUserContext.favoriteRecipe).toHaveBeenCalledWith('1', 'meals');
  });

  it('renders edit button when access is "personal"', async () => {
    const personalUserContext = { ...mockUserContext, access: 'personal' };
    const personalRecipe = { ...mockUserContext, ...mockRecipe, item: "meals" }

    DreamHostApi.getPersonalRecipe.mockResolvedValueOnce(personalRecipe);

    await act(async () => {
      render(
        <UserContext.Provider value={personalUserContext}>
          <MemoryRouter initialEntries={['/recipes/personal/meals/1']}>
            <Routes>
              <Route path="/recipes/:access/:item/:id" element={<RecipeDetails />} />
            </Routes>
          </MemoryRouter>
        </UserContext.Provider>
      );
    });

    const editButton = screen.getByRole('button', { name: /edit recipe/i });
    expect(editButton).toBeInTheDocument();
  });
});
