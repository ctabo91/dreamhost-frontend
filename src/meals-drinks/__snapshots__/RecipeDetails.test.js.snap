// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeDetails Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="RecipeDetails-container"
    style="background-image: url(test-thumbnail.jpg); background-size: cover; background-position: center; min-height: 100vh; width: 100%; padding-top: 60px; margin: 0px; position: relative;"
  >
    <div
      class="RecipeDetails-overlay"
    />
    <div
      class="RecipeDetails-content container"
    >
      <div
        class="row"
      >
        <h2
          class="RecipeDetails-title col text-center mt-4 "
        >
          TEST RECIPE
        </h2>
      </div>
      <div
        class="row"
      >
        <h3
          class="RecipeDetails-category col text-center"
        >
          Test Area Test Category Dish
        </h3>
      </div>
      <div>
        <div
          class="row mt-4 d-flex justify-content-around"
        >
          <div
            class="col-6"
          >
            <h4
              class="heading-title"
            >
              Instructions
            </h4>
            <ol
              class="RecipeDetails-instructions"
            >
              <li
                class="RecipeDetails-list-item"
              >
                Test instruction 
              </li>
              <li
                class="RecipeDetails-list-item"
              >
                Test instruction 2.
              </li>
            </ol>
          </div>
          <div
            class="col-1 split-line"
          />
          <div
            class="col-4"
          >
            <h4
              class="heading-title"
            >
              Ingredients
            </h4>
            <ul
              class="RecipeDetails-ingredients"
            >
              <li
                class="RecipeDetails-list-item"
              >
                Ingredient 1
              </li>
              <li
                class="RecipeDetails-list-item"
              >
                Ingredient 2
              </li>
            </ul>
          </div>
        </div>
        <button
          class="btn btn-favorite font-weight-bold mt-4 "
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-heart "
            data-icon="heart"
            data-prefix="far"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M225.8 468.2l-2.5-2.3L48.1 303.2C17.4 274.7 0 234.7 0 192.8v-3.3c0-70.4 50-130.8 119.2-144C158.6 37.9 198.9 47 231 69.6c9 6.4 17.4 13.8 25 22.3c4.2-4.8 8.7-9.2 13.5-13.3c3.7-3.2 7.5-6.2 11.5-9c0 0 0 0 0 0C313.1 47 353.4 37.9 392.8 45.4C462 58.6 512 119.1 512 189.5v3.3c0 41.9-17.4 81.9-48.1 110.4L288.7 465.9l-2.5 2.3c-8.2 7.6-19 11.9-30.2 11.9s-22-4.2-30.2-11.9zM239.1 145c-.4-.3-.7-.7-1-1.1l-17.8-20c0 0-.1-.1-.1-.1c0 0 0 0 0 0c-23.1-25.9-58-37.7-92-31.2C81.6 101.5 48 142.1 48 189.5v3.3c0 28.5 11.9 55.8 32.8 75.2L256 430.7 431.2 268c20.9-19.4 32.8-46.7 32.8-75.2v-3.3c0-47.3-33.6-88-80.1-96.9c-34-6.5-69 5.4-92 31.2c0 0 0 0-.1 .1s0 0-.1 .1l-17.8 20c-.3 .4-.7 .7-1 1.1c-4.5 4.5-10.6 7-16.9 7s-12.4-2.5-16.9-7z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
